### ISSUE TITLE
Cascade invocations should not suggest to merge a methodInvocation to another that is not void.

### ISSUE THREAD

**Issue by JPaulsen** [dart-lang/linter#602]
```dart
list = [];
list.asMap();
list.clear(); // OK because the fix would lose the value of list.asMap()
```

**Comment by a14n** [comment]
I don't understand why this should be special cased. There are a lot of method with non void return type. For instance:

```dart
list.remove('a');
list.add('a');

// vs.
list
  ..remove('a') // remove return a bool
  ..add('b');
```

The cascaded version seams good to me

**Comment by alexeieleusis** [comment]
The feedback we had internally (and the spirit of the operator) is to use it in builder like scenarios, for those the convention seems to be invoke a succession of methods with void return type (including setters). Also there is no easy way to programmatically determine when is OK to discard the return value from a non void method.

**Comment by a14n** [comment]
So perhaps 2 rules `cascade_invocations` and `cascade_invocations_on_declaration` that would apply only at   at the declaration. wdyt ?

**Comment by bwilkerson** [comment]
> The feedback we had internally (and the spirit of the operator) is to use it in builder like scenarios and fluent, for those the convention seems to be invoke a succession of methods with void return type (including setters).

That might be the most common use case, but I don't think the cascade operator was intended to be limited to that use case. :-)

> Also there is no easy way to programmatically determine when is OK to discard the return value from a non void method.

You don't need to. Looking at the original example, it's clear that the user has already determined that it's OK to discard the return value from `asMap`. Just trust the user. :-)

If we're getting feedback that users don't want that particular case to produce a lint, we need to understand why. Semantically the cascade and non-cascade versions of the example are equivalent, so there is no semantic argument for not creating a lint. What is the real objection?

**Comment by alexeieleusis** [comment]
I tried enabling it and cleaning up greentea and I was told what I wrote above, which I found to be a good use case. The problem was that there were too many lints in that category and we decided to disable the lint. I will try once more and more carefully document what was the problem.

**Comment by JPaulsen** [comment]
```
 it's clear that the user has already determined that it's OK to discard the return value from asMap. Just trust the user. :-)
```

That is true, but internally we had a case where something different happened and then we had to disable the rule, I would really like to know how was the case but it happened a long time ago so that was not easy job.

Anyway, the change in the implementation is so easy in the case we have to consider or not this issue, so I will be waiting for you to decide it, if something changes, it will be fixed quickly.

**Comment by bwilkerson** [comment]
Not knowing why we felt that we had to disable the rule, I don't have enough context to make a recommendation.

**Comment by JPaulsen** [comment]
Because of the feedbacks we are going to close this issue and modify the pull request

**Comment by pq** [comment]
@alexeieleusis : is this safe to close?

**Comment by alexeieleusis** [comment]
Definitely, once I go through our code base trying to enable this rule I'll create a new one if needed.
